### PR TITLE
Adapter methods: Serialisation to and from ladybug objects

### LINKED ISSUE #136: Refactor serialisation of python inputs/outputs
#### Description:

Move serialisation into an adapter so that their methods are hidden from the user.
- Create methods that convert BHoM objects to python readable json files.
- Read methods for converting from python json file output to BHoM objects.

### PR DESCRIPTION
**Pull request by Tom-Kingstone**
### Issues addressed by this PR

Closes #136 

Added the LadybugTools Adapter, with Pull and Push actions, that both use a `LadybugConfig` action config with a `FileSettings` property for a json file.
#### Pull:
This reads the contents of the json file, uses the BHoM serialiser to convert it into a dictionary (or custom object, depending on the layout of the json), checks against the `"type"` key in the dictionary and tries to convert the deserialised dictionary into a `LadybugTools` object of that type.
#### Push:
This takes in a `LadybugTools` object, and tries to convert it into a json serialised dictionary that can be read and deserialised into a ladybug readable python dictionary.

There were also some changes made to objects in the oM that might need handling in `versioning_70.json` and the versioning toolkit.

### Test files

[LBT_Tests.zip](https://github.com/BHoM/LadybugTools_Toolkit/files/13440534/LBT_Tests.zip)
Run this script using the serialised objects here:
[Serialised objects.zip](https://github.com/BHoM/LadybugTools_Toolkit/files/13440566/Serialised.objects.zip)

### CODE REVIEW

**Comment by Tom-Kingstone** [comment]
@BHoMBot check compliance

File: LadybugTools_Adapter/AdapterActions/Push.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+using System.Linq;
+using System.Text;
+
+namespace BH.Adapter.LadybugTools
+{
+    public partial class LadybugToolsAdapter : BHoMAdapter
+    {
+        public override List<object> Push(IEnumerable<object> objects, string tag = "", PushType pushType = PushType.AdapterDefault, ActionConfig actionConfig = null)
+        {
+            if (actionConfig is null)
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/AdapterActions/Push.cs:40]:
```suggestion
            if (actionConfig == null)
```

You're not in Python land here @Tom-Kingstone 😉 

I would also recommend trying to cast the config to the LadybugTools_Toolkit config and checking that can be done successfully and returning an error here if it's not, before you get to the `CreateLadybug` line.

File: LadybugTools_Adapter/CRUD/CreateLadybug.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+﻿using BH.oM.Adapter;
+using BH.oM.LadybugTools;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/CRUD/CreateLadybug.cs:31]:
```suggestion
```

File: LadybugTools_Adapter/CRUD/CreateLadybug.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    public partial class LadybugToolsAdapter : BHoMAdapter
+    {
+        public static bool CreateLadybug(List<ILadybugTools> objects, ActionConfig actionConfig = null)
+        {
+            LadybugConfig config = actionConfig as LadybugConfig;
+            if (config is null)
+            {
+                BH.Engine.Base.Compute.RecordError("Please input a valid LadybugConfig.");
+                return false;
+            }
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/CRUD/CreateLadybug.cs:43]: Move this check to the Push instead of handling it here.

File: LadybugTools_Adapter/CRUD/Read.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    public partial class LadybugToolsAdapter : BHoMAdapter
+    {
+        protected override IEnumerable<IBHoMObject> IRead(Type type, IList indices = null, ActionConfig actionConfig = null)
+        {
+            LadybugConfig config = actionConfig as LadybugConfig;
+            if (config == null)
+            {
+                BH.Engine.Base.Compute.RecordError($"The type of actionConfig provided: {actionConfig.GetType().FullName} is not valid for this adapter. Please provide a valid LadybugConfig actionConfig.");
+                return new List<IBHoMObject>();
+            }
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/CRUD/Read.cs:44]: Move this to the Pull.

File: LadybugTools_Adapter/CRUD/Read.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                BH.Engine.Base.Compute.RecordError($"The file at {config.JsonFile.GetFullFileName()} does not exist to pull from.");
+                return new List<IBHoMObject>();
+            }
+
+            List<IBHoMObject> rtnObjs = new List<IBHoMObject>();
+
+            rtnObjs.Add(config.JsonFile.ToBHoM());
+
+            return rtnObjs;
+
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/CRUD/Read.cs:63]: If you want it as different lines for debugging purposes then that's fine, but get rid of the blank lines between.

Otherwise, go with `return config.JsonFile.ToBHoM();` for a single line operation.

File: LadybugTools_Adapter/Convert/FromBHoM.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            File.WriteAllText(file.GetFullFileName(), json);
+            List<Event> events = BH.Engine.Base.Query.CurrentEvents();
+            return events.Count == 0;
+        }
+
+        public static string ICustomify(this ILadybugTools LBTObject)
+        {
+            if (LBTObject == null)
+            {
+                BH.Engine.Base.Compute.RecordError("Input object is null.");
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/FromBHoM.cs:50]: Update this error message to give more information to the user on what was trying to be done so they can best rectify the issue.

File: LadybugTools_Adapter/Convert/FromBHoM.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+                return null;
+            }
+            return Jsonify(LBTObject as dynamic);
+        }
+
+        private static string Jsonify(this oM.LadybugTools.AnalysisPeriod analysisPeriod)
+        {
+            return FromAnalysisPeriod(analysisPeriod).ToJson();
+        }
+        private static string Jsonify(this oM.LadybugTools.DataType dataType)
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/FromBHoM.cs:60]:
```suggestion

        private static string Jsonify(this oM.LadybugTools.DataType dataType)
```

File: LadybugTools_Adapter/Convert/MetaData/DataType.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace BH.Adapter.LadybugTools
+{
+    public static partial class Convert
+    {
+        public static BH.oM.LadybugTools.DataType ToDataType(Dictionary<string, object> oldObject)
+        {
+            string BaseUnit;
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/MetaData/DataType.cs:34]: Lowercase `b` please on local variable name.

**Comment by Tom-Kingstone** [comment]
@BHoMBot fix copyright headers ref. 18933886803

**Comment by Tom-Kingstone** [comment]
@BHoMBot fix project file ref. 18933872438

**Comment by Tom-Kingstone** [comment]
@BHoMBot check compliance

File: LadybugTools_oM/Collections/Collection/DoublesCollection.cs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using BH.oM.Base;
+
+namespace BH.oM.LadybugTools
+{
+    public class DoublesCollection : BHoMObject, ICollection
+    {
+        [Description("A list of doubles.")]
+        public virtual List<double> Values { get; set; } = null;
```
Comment (FraserGreenroyd) [inline LadybugTools_oM/Collections/Collection/DoublesCollection.cs:34]: Why is this here? This hardly seems beneficial to the oM? Is this meant to help the adapter work? If so, this should go into an `Objects` folder within the adapter rather than the public oM.

File: LadybugTools_oM/Collections/Collection/ICollection.cs [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+using System.ComponentModel;
+using System.Collections.Generic;
+using System.Linq;
+using BH.oM.Base;
+
+namespace BH.oM.LadybugTools
+{
+    [Description("An interface for different types of collections.")]
+    public interface ICollection : IBHoMObject
```
Comment (FraserGreenroyd) [inline LadybugTools_oM/Collections/Collection/ICollection.cs:32]: Same here

File: LadybugTools_oM/Collections/Collection/StringsCollection.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using BH.oM.Base;
+
+namespace BH.oM.LadybugTools
+{
+    public class StringsCollection : BHoMObject, ICollection
+    {
+        [Description("A list of strings.")]
+        public virtual List<string> Values { get; set; } = null;
```
Comment (FraserGreenroyd) [inline LadybugTools_oM/Collections/Collection/StringsCollection.cs:34]: Same here

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
On the converts, everywhere we're doing a cast (such as `(double)value;`) we should wrap into a try-catch and provide the data separately to returning the object, just to protect against unexpected crashes from potentially dodgy data.

Otherwise this is a good start @Tom-Kingstone , great work 😄

File: LadybugTools_oM/Collections/Collection/DoublesCollection.cs (hunk 9 above)
Comment (Tom-Kingstone) [reply LadybugTools_oM/Collections/Collection/DoublesCollection.cs:34]: After discussion, `HourlyContinuousCollection` will use List<string> for collections, which while it means that anyone using the lists as doubles must parse the strings to doubles, it also removes the Doubles/Strings Collection objects so users will be less confused.

**Comment by Tom-Kingstone** [comment]
@BHoMBot check compliance

**Comment by Tom-Kingstone** [comment]
@BHoMBot check core

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: installer, null-handling, serialisation, unit-tests, ready-to-merge

File: LadybugTools_Adapter/AdapterActions/Pull.cs [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+using System.Text;
+
+namespace BH.Adapter.LadybugTools
+{
+    public partial class LadybugToolsAdapter : BHoMAdapter
+    {
+        public override IEnumerable<object> Pull(IRequest request, PullType pullType = PullType.AdapterDefault, ActionConfig actionConfig = null) 
+        {
+            LadybugConfig config = actionConfig as LadybugConfig;
+            if (config == null)
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/AdapterActions/Pull.cs:39]: You should still check that `actionConfig != null` first, otherwise if it is, it won't cast to `LadybugConfig` but it also won't exist for `actionConfig.GetType()` to work either.

File: LadybugTools_Adapter/AdapterActions/Pull.cs [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+                return new List<IBHoMObject>();
+            }
+
+            if (!System.IO.File.Exists(config.JsonFile.GetFullFileName()))
+            {
+                BH.Engine.Base.Compute.RecordError($"The file at {config.JsonFile.GetFullFileName()} does not exist to pull from.");
+                return new List<IBHoMObject>();
+            }
+
+
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/AdapterActions/Pull.cs:57]:
```suggestion
```

File: LadybugTools_Adapter/AdapterActions/Push.cs [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+            if (objects.Count() == 0)
+            {
+                BH.Engine.Base.Compute.RecordError("Please input a valid LadybugTools object.");
+                return new List<object>();
+            }
+
+            if (objects.Count() > 1)
+                BH.Engine.Base.Compute.RecordWarning("The LadybugToolsAdapter does not currently support pushing multiple objects to one file, only the first object will be saved.");
+
+
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/AdapterActions/Push.cs:56]:
```suggestion
```

File: LadybugTools_Adapter/CRUD/CreateLadybug.cs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+using System.Linq;
+using System.Text;
+
+namespace BH.Adapter.LadybugTools
+{
+    public partial class LadybugToolsAdapter : BHoMAdapter
+    {
+        public static bool CreateLadybug(List<ILadybugTools> objects, LadybugConfig config = null)
+        {
+
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/CRUD/CreateLadybug.cs:37]:
```suggestion
```

File: LadybugTools_Adapter/CRUD/CreateLadybug.cs [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+namespace BH.Adapter.LadybugTools
+{
+    public partial class LadybugToolsAdapter : BHoMAdapter
+    {
+        public static bool CreateLadybug(List<ILadybugTools> objects, LadybugConfig config = null)
+        {
+            
+            if (!Convert.FromBHoM(objects[0], config.JsonFile))
+            {
+                BH.Engine.Base.Compute.RecordError("An error occurred during conversion to json:");
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/CRUD/CreateLadybug.cs:40]:
```suggestion
                BH.Engine.Base.Compute.RecordError("An error occurred during conversion to Ladybug JSON.");
```

File: LadybugTools_Adapter/CRUD/Read.cs [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+namespace BH.Adapter.LadybugTools
+{
+    public partial class LadybugToolsAdapter : BHoMAdapter
+    {
+        protected override IEnumerable<IBHoMObject> IRead(Type type, IList indices = null, ActionConfig actionConfig = null)
+        {
+            LadybugConfig config = actionConfig as LadybugConfig;
+            
+
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/CRUD/Read.cs:41]:
```suggestion
```

File: LadybugTools_Adapter/Convert/Collections/EPW.cs [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+                    collections.Add(ToHourlyContinuousCollection((collection as CustomObject).CustomData));
+                else
+                    collections.Add(ToHourlyContinuousCollection(collection as Dictionary<string, object>));
+            }
+            EPW epw = new EPW()
+            {
+                Location = ToLocation(oldObject["location"] as Dictionary<string, object>),
+                DataCollections = collections
+            };
+            try
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/Collections/EPW.cs:57]:
```suggestion

            try
```

File: LadybugTools_Adapter/Convert/Collections/HourlyContinuousCollection.cs [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+namespace BH.Adapter.LadybugTools
+{
+    public static partial class Convert
+    {
+        public static BH.oM.LadybugTools.HourlyContinuousCollection ToHourlyContinuousCollection(Dictionary<string, object> oldObject)
+        {
+            List<string> hourlyValues = new List<string>();
+            if (oldObject["header"].GetType() == typeof(CustomObject))
+                oldObject["header"] = (oldObject["header"] as CustomObject).CustomData;
+            try
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/Collections/HourlyContinuousCollection.cs:41]:
```suggestion

            try
```

File: LadybugTools_Adapter/Convert/Collections/HourlyContinuousCollection.cs [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+            {
+               hourlyValues = (oldObject["values"] as List<object>).Select(x => x.ToString()).ToList();
+            }
+            catch (Exception ex)
+            {
+                BH.Engine.Base.Compute.RecordError($"An error occurred when converting the values in a collection. Returning an empty collection: \n The error: {ex}");
+            }
+            return new oM.LadybugTools.HourlyContinuousCollection()
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/Collections/HourlyContinuousCollection.cs:49]:
```suggestion

            return new oM.LadybugTools.HourlyContinuousCollection()
```

File: LadybugTools_Adapter/Convert/Constructions/EnergyMaterial.cs [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+                        Name = (string)oldObject["identifier"],
+                        Thickness = (double)oldObject["thickness"],
+                        Conductivity = (double)oldObject["conductivity"],
+                        Density = (double)oldObject["density"],
+                        SpecificHeat = (double)oldObject["specific_heat"],
+                        Roughness = roughness,
+                        ThermalAbsorptance = (double)oldObject["thermal_absorptance"],
+                        SolarAbsorptance = (double)oldObject["solar_absorptance"],
+                        VisibleAbsorptance = (double)oldObject["visible_absorptance"]
+                    };
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/Constructions/EnergyMaterial.cs:49]: It will expand the method somewhat, but I would have these being all their own try-catches, and a suitable default used if they can't convert the type successfully. That way, we might get some of the data even if we can't get all of it

File: LadybugTools_Adapter/Convert/Constructions/EnergyMaterialVegetation.cs [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        public static BH.oM.LadybugTools.EnergyMaterialVegetation ToEnergyMaterialVegetation(Dictionary<string, object> oldObject)
+        {
+            if (Enum.TryParse((string)oldObject["roughness"], out Roughness roughness))
+            {
+                try
+                {
+                    return new oM.LadybugTools.EnergyMaterialVegetation()
+                    {
+                        Name = (string)oldObject["identifier"],
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/Constructions/EnergyMaterialVegetation.cs:40]: Same here

File: LadybugTools_Adapter/Convert/MetaData/AnalysisPeriod.cs [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+{
+    public static partial class Convert
+    {
+        public static BH.oM.LadybugTools.AnalysisPeriod ToAnalysisPeriod(Dictionary<string, object> oldObject)
+        {
+            try
+            {
+                return new oM.LadybugTools.AnalysisPeriod()
+                {
+                    StartMonth = (int)oldObject["st_month"],
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/MetaData/AnalysisPeriod.cs:39]: Same here

File: LadybugTools_Adapter/Convert/MetaData/DataType.cs [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+    public static partial class Convert
+    {
+        public static BH.oM.LadybugTools.DataType ToDataType(Dictionary<string, object> oldObject)
+        {
+            string baseUnit;
+            if (oldObject.ContainsKey("base_unit"))
+                baseUnit = (string)oldObject["base_unit"];
+            else
+                baseUnit = "";
+            try
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/MetaData/DataType.cs:40]:
```suggestion

            try
```

File: LadybugTools_Adapter/Convert/MetaData/Header.cs [hunk 25]
```diff
@@ -0,0 +1,75 @@
...
+
+            if (oldObject["metadata"].GetType() == typeof(CustomObject))
+                oldObject["metadata"] = (oldObject["metadata"] as CustomObject).CustomData;
+
+            Header header = new Header()
+            {
+                DataType = ToDataType(oldObject["data_type"] as Dictionary<string, object>),
+                AnalysisPeriod = ToAnalysisPeriod(oldObject["analysis_period"] as Dictionary<string, object>)
+            };
+            try
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/MetaData/Header.cs:50]:
```suggestion

            try
```

File: LadybugTools_Adapter/Convert/MetaData/Location.cs [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+{
+    public static partial class Convert
+    {
+        public static BH.oM.LadybugTools.Location ToLocation(Dictionary<string, object> oldObject)
+        {
+            try
+            {
+                return new oM.LadybugTools.Location()
+                {
+                    City = (string)oldObject["city"],
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/MetaData/Location.cs:38]: Same here

File: LadybugTools_Adapter/Convert/ToBHoM.cs [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+                lbtObject = (obj as CustomObject).CustomData;
+            }
+            else if (obj.GetType() == typeof(Dictionary<string, object>))
+            {
+                BH.Engine.Base.Compute.RecordWarning("The object was not deserialised as a CustomObject, are you sure that this file came from a LadybugTools Python object? \n Trying to cast to Dictionary...");
+                lbtObject = obj as Dictionary<string, object>;
+            }
+            else
+            {
+                BH.Engine.Base.Compute.RecordWarning($"The json given already deserialises to a BHoM object of type: {obj.GetType().FullName} - please use the BHoM Engine serialiser to deserialise this object.");
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/ToBHoM.cs:56]: Why can't we do this for the user so they have a streamlined experience?

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by Tom-Kingstone** [comment]
@BHoMBot check compliance
@BHoMBot check core

**Comment by Tom-Kingstone** [comment]
@BHoMBot check required

**Comment by Tom-Kingstone** [comment]
@BHoMBot check compliance
@BHoMBot check required

File: LadybugTools_Adapter/Convert/Constructions/EnergyMaterialVegetation.cs [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+            {
+                specificHeat = (double)oldObject["specific_heat"];
+            }
+            catch (Exception ex)
+            {
+                BH.Engine.Base.Compute.RecordError($"An error occurred when reading the specific heat of the EnergyMaterial. returning specific heat as default ({specificHeat}).\n The error: {ex}");
+            }
+
+
+
```
Comment (FraserGreenroyd) [inline LadybugTools_Adapter/Convert/Constructions/EnergyMaterialVegetation.cs:95]:
```suggestion
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Review by FraserGreenroyd — APPROVED** [review]
Good work @Tom-Kingstone nice to get this over the line and deployed for alpha testing with the users

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning
@BHoMBot check compliance
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: unit-tests

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge
